tests/tape/tapes/test_qnode.py

```diff
@@ -329,6 +329,59 @@ def func(x, y):
         assert qn.qtape.operations == [op1, op2, op3]
         assert qn.qtape.measurements == [m1, m2]
 
+
+    def test_make_drawing(self):
+        """Test circuit drawing"""
+        from pennylane import numpy as anp
+
+        x = anp.array(0.1, requires_grad=True)
+        y = anp.array([0.2, 0.3], requires_grad=True)
+        z = anp.array(0.4, requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev, interface="autograd")
+        def circuit(p1, p2=y, **kwargs):
+            qml.RX(p1, wires=0)
+            qml.RY(p2[0] * p2[1], wires=1)
+            qml.RX(kwargs["p3"], wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        result = make_drawing(circuit)(p1=x, p3=z)
+        expected = """\
+ 0: ──RX(0.1)───RX(0.4)──╭C──╭┤ ⟨Z ⊗ X⟩ 
+ 1: ──RY(0.06)───────────╰X──╰┤ ⟨Z ⊗ X⟩ 
+"""
+
+        assert result == expected
+
+    def test_make_drawing_ascii(self):
+        """Test circuit drawing when using ASCII characters"""
+        from pennylane import numpy as anp
+
+        x = anp.array(0.1, requires_grad=True)
+        y = anp.array([0.2, 0.3], requires_grad=True)
+        z = anp.array(0.4, requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev, interface="autograd")
+        def circuit(p1, p2=y, **kwargs):
+            qml.RX(p1, wires=0)
+            qml.RY(p2[0] * p2[1], wires=1)
+            qml.RX(kwargs["p3"], wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        result = make_drawing(circuit, charset="ascii")(p1=x, p3=z)
+        expected = """\
+ 0: --RX(0.1)---RX(0.4)--+C--+| <Z @ X> 
+ 1: --RY(0.06)-----------+X--+| <Z @ X> 
+"""
+
+        assert result == expected

```

Comment:
Not sure myself, but would it be worth adding a test where the qnode gets evaluated and/or parameters set before drawing? Just to catch potential side effects.

Also, if I evaluate the qnode before drawing, can I avoid passing arguments to the drawing functions, since it can just use the previous ones? For example 

```python
        circuit(p1=x, p3=z)
        result = make_drawing(circuit, charset="ascii")
```

Reply:
I really want to avoid side effects entirely whenever possible now. I don't think it's a good idea to also have `drawing` depend on whether the `qnode` was called already or not and return different types in each case (function vs string).

I think the compromise here is to keep the `qnode.draw()` method that just draws with the last known argument.